### DISCUSSION TITLE
Working with TypeScript and the ubiquitous client-passing

### DISCUSSION THREAD

**Discussion by Nicholaiii** [Snazzah/slash-create#204] [Ideas]
In the FAQ you clearly state you don't support passing the client through, but I think you missed a great feature of require-all; the `resolve` option, that parses the contents of the module before handing it back to your library. Using this option, we can export functions that takes the parameters we need to inject, and return a class that has this in its closure.
Functional example:

```ts
creator.registerCommandsIn({
  dirname: join(__dirname, 'commands'),
  filter: /^([^.].*)\.(js|ts)$/,
  resolve: function ({ default: makeCommand } : { default: (client: Client) => SlashCommand }) {
    return makeCommand(client)
  }
})
```

```ts
/* commands/test.ts */

export default function (client: Client) {
  class TestCommand extends SlashCommand {
    constructor (creator: SlashCreator, options: SlashCommandOptions) {
      super(creator, {
        name: 'test',
        description: 'Testing',
      })
    }
    
    async run (ctx: CommandContext) {
      /* client available here */
      return 'Hello world'
    }
  }

  return TestCommand
}

```

I think this approach is worth adopting and using in the guide, and not just for TypeScript.

**Comment by Snazzah** [comment]
Nice Idea, definitely better than the alternatives. However, I think i'll probably end up allowing people to set a client into the creator so commands can access it easily. This solution also will have the problem of not being able to use it in [slash-up](https://github.com/Snazzah/slash-up). Although slash-create is mainly focused on HTTP only bots, this won't be much of a hassle to do. The only thing that might be annoying is TypeScript devs having to declare the type for every command. Not to mention, I was thinking of replacing require-all since it's not more useful than providing a path.

**Reply by Nicholaiii** [reply]
Yeah, I understand this. A commonly used pattern is making the class take a generic type for a meta object:
```ts
class SlashCommand <M extends Record<any, any>> {
  meta: M
  /* ...rest */
}
```
This keeps it system agnostic but highly compatible, even in a (non-breaking) backwards manner, with a well-defined approach for adding custom values.

require-all could easily be replaced with `readdir` and `import()`
